### ISSUE TITLE
What do you think about using events?

### ISSUE THREAD

**Issue by TheoD02** [jolicode/castor#217]
What do you think about using events in castor ?

This could allow flexibility and for example run a specific command, for commands starting with `keycloak:` for the example?

This make sure keycloak is started
```php
register_event_listener(
    BeforeExecuteTaskEvent::class,
    function (BeforeExecuteTaskEvent $event): void {
        if (u($event->getCommand()->getName())->startsWith('keycloak:')) {
            // Do something before the command is executed
           $isKeycloakUp = (bool) (capture("docker ps -a -q -f name=my-keycloak-container", allowFailure: true));
           if ($isKeycloakUp === false) {
                run('docker compose up -d keycloak', path: 'path/to/keycloak');
            }
        }
    }
);
```

**Comment by lyrixx** [comment]
I already thought about something like this.

The way you wrote it could not work, since we don't execute anything from loaded file.
I think we could use an Attribute for this. Like it's done for a task or a context.

```php
#[AsListener(BeforeTaskEvent::class)
on_foobar($event) {
}
```


WDYT?

**Comment by TheoD02** [comment]
> I already thought about something like this.
> 
> The way you wrote it could not work, since we don't execute anything from loaded file. I think we could use an Attribute for this. Like it's done for a task or a context.
> 
> ```
> #[AsListener(BeforeTaskEvent::class)
> on_foobar($event) {
> }
> ```
> 
> WDYT?

Yes indeed, it works, I did a POC on the subject quickly to see the possibilities, but in my example, the listener registration must be first before functions in the castor.php. You're right, this could lead to unexpected behavior ^^

The logic with the attribute seems the closest to what already exists as in Symfony, and the best 😃 

I can create the PR if needed 😄

**Comment by lyrixx** [comment]
> I can create the PR if needed 😄

Let's go :)

**Comment by lyrixx** [comment]
Fixed in #221
